pkg/certs/rotate.go

```diff
@@ -0,0 +1,219 @@
+	}
+
+	// In standalone there is no host secret so we skip updating the secret.
+	if vConfig.ControlPlane.Standalone.Enabled {
+		return nil
+	}
+
+	// Update the secret so in case of a restart without persistence we don't loose data.
+	return updateSecret(ctx, vConfig.ControlPlaneNamespace, CertSecretName(vConfig.Name), constants.PKIDir, vConfig.ControlPlaneClient)
+}
+
+func backupDirectory(src, dst string) error {
+	if err := os.MkdirAll(dst, 0755); err != nil {
+		return fmt.Errorf("creating destination directory: %w", err)
+	}
+
+	return filepath.WalkDir(src, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		relPath, err := filepath.Rel(src, path)
+		if err != nil {
+			return err
+		}
+		dstPath := filepath.Join(dst, relPath)
+
+		if d.IsDir() {
+			return os.MkdirAll(dstPath, 0755)
+		}
+
+		return copyFile(path, dstPath)
+	})
+}
+
+func copyFile(src, dst string) error {
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	defer srcFile.Close()
+
+	dstFile, err := os.Create(dst)
+	if err != nil {
+		return err
+	}
+	defer dstFile.Close()
+
+	_, err = io.Copy(dstFile, srcFile)
+	if err != nil {
+		return err
+	}
+
+	srcInfo, err := srcFile.Stat()
+	if err != nil {
+		return err
+	}
+
+	return os.Chmod(dst, srcInfo.Mode())
+}
+
+type excludeFunc func(name string) bool
+
+func removeFiles(path string, excludeFuncs ...excludeFunc) error {
+	return filepath.WalkDir(path, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if d.IsDir() {
+			return nil
+		}
+
+		for _, shouldExclude := range excludeFuncs {
+			if shouldExclude(d.Name()) {
+				return nil
+			}
+		}
+
+		if err := os.Remove(path); err != nil {
+			return fmt.Errorf("removing %s: %w", path, err)
+		}
+
+		return nil
+	})
+}
+
+func excludeCAFiles(name string) bool {
+	if name == CACertName {
+		return true
+	}
+	if name == CAKeyName {
+		return true
+	}
+	if strings.HasSuffix(name, fmt.Sprintf("-%s", CACertName)) {
+		return true
+	}
+	if strings.HasSuffix(name, fmt.Sprintf("-%s", CAKeyName)) {
+		return true
+	}
+	return false
+}
+
+func excludeSAFiles(name string) bool {
+	if name == ServiceAccountPublicKeyName {
+		return true
+	}
+	if name == ServiceAccountPrivateKeyName {
+		return true
+	}
+	return false
+}
+
+func updateSecret(ctx context.Context, secretNamespace, secretName, pkiPath string, client kubernetes.Interface) error {
+	secret, err := client.CoreV1().Secrets(secretNamespace).Get(ctx, secretName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("getting cert secret %s: %w", secretName, err)
+	}
+
+	data := map[string][]byte{}
+	for k, v := range certMap {
+		d, err := os.ReadFile(filepath.Join(pkiPath, k))
+		if err != nil {
+			return fmt.Errorf("reading file %s: %w", k, err)
+		}
+
+		data[v] = d
+	}
+	secret.Data = data
+
+	_, err = client.CoreV1().Secrets(secretNamespace).Update(ctx, secret, metav1.UpdateOptions{})
```

Comment:
nit: prefer a [patch](https://github.com/loft-sh/vcluster/blob/main/pkg/k8s/migrate.go#L203-L208) to an update.